File: 0169-majority-element/0169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        int count = 0;
        int candidate = 0;
        int N = nums.length;


        for (int i=0; i<N; i++){
            if (count == 0){
                candidate = nums[i];
                count = 1;
            }else if(candidate == nums[i]){
                count += 1;
            }else{
                count -= 1;
            }
        }

        int count1 = 0;
        for (int i = 0; i< N; i++){
            if (nums[i] == candidate){
                count1++;
            }
        }

        if(count1 > (N/2)){
            return candidate;
        }

        return -1;
    }
}

File: 0169-majority-element/0169-majority-element.py
class Solution:
    #Brute Force approach using HashMap TC-O(n) and SC-O(n)
    # def majorityElement(self, nums: List[int]) -> int:
    #     count = defaultdict(int)
    #     candidate, maxCount = 0, 0

    #     for num in nums:
    #         count[num] += 1
    #         if maxCount < count[num]:
    #             candidate = num
    #             maxCount = count[num]
    #     return candidate
    def majorityElement(self, nums: List[int]) -> int:
        #Optimal Approach Using Moore Vooting Algo TC-O(n) and SC-O(1)
        count, candidate = 0, None

        for num in nums:
            if count == 0:
                candidate = num
                count = 1
            elif candidate == num:
                count += 1
            else:
                count -= 1
        
        if nums.count(candidate) > len(nums) // 2:
            return candidate
        return None
        